linera-views/src/views/bucket_log_view.rs

```diff
@@ -0,0 +1,492 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(with_metrics)]
+use std::sync::LazyLock;
+use std::{
+    collections::{BTreeSet, HashMap},
+    ops::{Bound, Range, RangeBounds},
+};
+
+use async_trait::async_trait;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+#[cfg(with_metrics)]
+use {
+    linera_base::prometheus_util::{
+        exponential_bucket_latencies, register_histogram_vec, MeasureLatency,
+    },
+    prometheus::HistogramVec,
+};
+
+use crate::{
+    batch::Batch,
+    common::{from_bytes_option_or_default, HasherOutput},
+    context::Context,
+    hashable_wrapper::WrappedHashableContainerView,
+    views::{ClonableView, HashableView, Hasher, View, ViewError, MIN_VIEW_TAG},
+};
+
+#[cfg(with_metrics)]
+/// The runtime of hash computation
+static BUCKET_LOG_VIEW_HASH_RUNTIME: LazyLock<HistogramVec> = LazyLock::new(|| {
+    register_histogram_vec(
+        "bucket_log_view_hash_runtime",
+        "BucketLogView hash runtime",
+        &[],
+        exponential_bucket_latencies(5.0),
+    )
+});
+
+/// Key tags to create the sub-keys of a `LogView` on top of the base key.
+#[repr(u8)]
+enum KeyTag {
+    /// Prefix for the storing of the variable `stored_count`.
+    StoredSizes = MIN_VIEW_TAG,
+    /// Prefix for the indices of the log.
+    Index,
+}
+
+/// The `StoredIndices` contains the description of the stored buckets.
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+struct StoredSizes {
+    /// The stored buckets with the first index being the size (at most N) and the
+    /// second one is the index in the storage. If the index is 0 then it corresponds
+    /// with the first value (entry `KeyTag::Front`), otherwise to the keys with
+    /// prefix `KeyTag::Index`.
+    sizes: Vec<usize>,
+}
+
+#[derive(Clone, Debug)]
+enum Bucket<T> {
+    Loaded { data: Vec<T> },
+    NotLoaded { length: usize },
+}
+
+impl<T> Bucket<T> {
+    fn len(&self) -> usize {
+        match self {
+            Bucket::Loaded { data } => data.len(),
+            Bucket::NotLoaded { length } => *length,
+        }
+    }
+}
+
+/// A view that supports logging values of type `T`.
+#[derive(Debug)]
+pub struct BucketLogView<C, T, const N: usize> {
+    context: C,
+    delete_storage_first: bool,
+    stored_data: Vec<Bucket<T>>,
```

Comment:
(I wonder at what length this becomes less efficient than the regular log view for e.g. just accessing the last element. We now always have to load the whole list of bucket sizes.)